File: java/Problem006_ZigZagConversion.java
// https://leetcode.com/problems/zigzag-conversion

class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1 || s.length() <= numRows) return s;

        StringBuilder[] rows = new StringBuilder[numRows];

        for (int i=0; i<numRows; i++) {
            rows[i] = new StringBuilder();
        }
        int increment = -1;
        int row = 0;

        for (char c : s.toCharArray()) {
            rows[row].append(c);

            if (row == 0 || row == (numRows-1)) {
                increment *= -1;
            }
            row += increment;
        }

        StringBuilder result = new StringBuilder();
        for (int i=0; i<numRows; i++) {
            result.append(rows[i]);
        }

        return result.toString();

    }
}

File: python/problem_006_zigzag_conversion.py
# https://leetcode.com/problems/zigzag-conversion

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or len(s) <= numRows:
            return s

        rows = [[] for _ in range(numRows)]
        increment = 1
        row = 0

        for c in s:
            rows[row].append(c)

            if row == 0:
                increment = 1
            elif row == numRows-1:
                increment = -1

            row += increment

        return "".join("".join(r) for r in rows)
        
